compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -852,6 +853,36 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         let mut suggested =
                             self.suggest_dereferences(&obligation, &mut err, trait_predicate);
                         suggested |= self.suggest_fn_call(&obligation, &mut err, trait_predicate);
+                        let impl_candidates = self.find_similar_impl_candidates(trait_predicate);
+                        suggested = if let &[cand] = &impl_candidates[..] {
+                            let cand = cand.trait_ref;
+                            if let (ty::FnPtr(_), ty::FnDef(fn_did, _)) =
+                                (cand.self_ty().kind(), trait_ref.self_ty().skip_binder().kind())
+                            {
+                                err.span_suggestion(
+                                    span,
```

Comment:
Let's just suggest the cast, without rewriting what the user wrote: it may be an alias `let my_func = example;` or `use example as my_func;` instead of the function's actual name `example`.

```suggestion
                                    span.shrink_to_hi(),
```